ladon.go

```diff
@@ -66,57 +80,78 @@ func (l *Ladon) DoPoliciesAllow(r *Request, policies []Policy) (err error) {
 
 	// Iterate through all policies
 	for _, p := range policies {
-		// Does the action match with one of the policies?
-		// This is the first check because usually actions are a superset of get|update|delete|set
-		// and thus match faster.
-		if pm, err := l.matcher().Matches(p, p.GetActions(), r.Action); err != nil {
-			return errors.WithStack(err)
-		} else if !pm {
-			// no, continue to next policy
-			continue
-		}
 
 		// Does the subject match with one of the policies?
 		// There are usually less subjects than resources which is why this is checked
 		// before checking for resources.
 		if sm, err := l.matcher().Matches(p, p.GetSubjects(), r.Subject); err != nil {
+			if l.Metric != nil {
+				go l.Metric.RequestProcessingError(*r, p, err)
+			}
 			return err
 		} else if !sm {
 			// no, continue to next policy
 			continue
 		}
 
-		// Does the resource match with one of the policies?
-		if rm, err := l.matcher().Matches(p, p.GetResources(), r.Resource); err != nil {
+		// Does the action match with one of the policies?
+		// This is the first check because usually actions are a superset of get|update|delete|set
+		// and thus match faster.
+		if pm, err := l.matcher().Matches(p, p.GetActions(), r.Action); err != nil {
+			if l.Metric != nil {
+				go l.Metric.RequestProcessingError(*r, p, err)
+			}
 			return errors.WithStack(err)
-		} else if !rm {
+		} else if !pm {
 			// no, continue to next policy
 			continue
 		}
 
-		// Are the policies conditions met?
-		// This is checked first because it usually has a small complexity.
-		if !l.passesConditions(p, r) {
+		// Does the resource match with one of the policies?
+		if rm, err := l.matcher().Matches(p, p.GetResources(), r.Resource); err != nil {
+			if l.Metric != nil {
+				go l.Metric.RequestProcessingError(*r, p, err)
+			}
+			return errors.WithStack(err)
+		} else if !rm {
 			// no, continue to next policy
 			continue
 		}
 
-		// Is the policies effect deny? If yes, this overrides all allow policies -> access denied.
+		// Is the policies effect deny? If yes, this overrides all allow policies
 		if !p.AllowAccess() {
 			deciders = append(deciders, p)
 			l.auditLogger().LogRejectedAccessRequest(r, policies, deciders)
+			if l.Metric != nil {
+				go l.Metric.RequestDeniedBy(*r, p)
+			}
 			return errors.WithStack(ErrRequestForcefullyDenied)
 		}
 
+		// Are the policies conditions met?
+		// This is checked first because it usually has a small complexity.
+		if !l.passesConditions(p, r) {
```

Comment:
This should move back to where it was - otherwise policies not matching the context can deny access.